src/envoy/http/authn/authn_utils.cc

```diff
@@ -89,37 +62,39 @@ bool AuthnUtils::ProcessJwtPayload(const std::string& payload_str,
   }
 
   *payload->mutable_raw_claims() = payload_str;
-  ::google::protobuf::Map< ::std::string, ::std::string>* claims =
-      payload->mutable_claims();
 
-  // Extract claims
-  json_obj->iterate(
-      [payload](const std::string& key, const Json::Object& obj) -> bool {
-        ::google::protobuf::Map< ::std::string, ::std::string>* claims =
-            payload->mutable_claims();
-        // In current implementation, only string objects are extracted into
-        // claims. If call obj.asJsonString(), will get "panic: not reached"
-        // from json_loader.cc.
-        try {
-          // Try as string, will throw execption if object type is not string.
-          (*claims)[key] = obj.asString();
-        } catch (Json::Exception& e) {
-        }
-        return true;
-      });
-  // Extract audience
-  // ExtractJwtAudience() should be called after claims are extracted.
-  ExtractJwtAudience(*json_obj, payload->claims(), payload);
-  // ExtractJwtGroups() should be called after claims are extracted.
-  ExtractJwtGroups(*json_obj, payload->claims(), payload);
+  auto claims = payload->mutable_claims()->mutable_fields();
+  // Extract claims as string lists
+  json_obj->iterate([json_obj, claims](const std::string& key,
+                                       const Json::Object&) -> bool {
+    // In current implementation, only string/string list objects are extracted
```

Comment:
Instead of looping over the JSON object manually, why not use `ParseJsonMessage()` to parse the whole json string to the `claim` field? So that you could simplify the ProcessJwtPayload() a lot.

Reply:
@yangminzhu The output of ParseJsonMessage() is of the type Message: 
Status ParseJsonMessage(const std::string& json, Message* output). Here the claims is of the type protobuf.Struct. Can you provide details of how to use ParseJsonMessage() to parse the json object into protobuf.Struct? One more thing, if the json object is parsed into the claims Struct as different types (e.g., strings, string lists, structs, and etc), how does the consumer of the claims interpret the type of a claim (since the type for a claim value is not only a string list anymore)?

Reply:
`Struct` is a subclass of `Message`, so it should be able to pass to `ParseJsonMessage()` directly.

I'm curious where are we using the `claims` field other than in `authn_utils.cc`? The type is changed and it's seems only used in this file directly?

@diemtvu WDYT?

Reply:
The consumers of the claims protobuf.Struct include the producers of dynamic metadata and tests in the Istio proxy repo. If the claims become a map from string key to protobuf.Struct value, every consumer of the claims needs to add some code to verify the type of the claim value before accessing a claim. In contrast, with current implementation, the contract between the producer of the claims and the consumers is that a claim value must be of the string list type and a consumer directly reads a claim value as a string list without extra code to verify the type of a claim.

Reply:
The `claims`'s type is already changed to `Struct` so why it matters about how you construct it? If the constructed `Struct` doesn't have a string or string list as your implicit contract here, the consumer still get the same value as today when using the `.string_value()` method on the `Struct`. So there shouldn't have any changes on consumer side once its stored in dynamic metadata.

Reply:
You can explore, but I do think ParseJsonMessage, even if it works with Struct, may has some limitation. For example, in our approach, we want to have consistent representation for string + string array. I'm ok with this custom conversion for now, and find a better utils later (or refactor them to a util function)

Reply:
Could you point me to one of the consumer side code that needs to change if you use `ProcessJwtPayload()`?

Reply:
The following is a few examples on the consumer side code that need to change if use ParseJsonMessage():
- In /usr/local/google/home/leitang/git/istio/proxy/src/envoy/utils/authn.cc: 
    payload->set_user(
        (*claims)["iss"].list_value().values().Get(0).string_value() + "/" +
        (*claims)["sub"].list_value().values().Get(0).string_value());
    payload->set_presenter(
        (*claims)["azp"].list_value().values().Get(0).string_value());

Reply:
Another example on the consumer side code that need to change if use ParseJsonMessage():
  // Copy audience to the audience in context.proto
  if (claims->find(kJwtAudienceKey) != claims->end()) {
    for (const auto& v : (*claims)[kJwtAudienceKey].list_value().values()) {
      payload->add_audiences(v.string_value());
    }
  }

Reply:
These are just right in this function, I'm trying to see if there are any consumer code outside this file and I will be supersized if it is consuming the `claim` field the same way as the examples, i.e. getting string directly from the list.

Anyway, I'm ok if you would like to go with the current approach, we can discuss more and refactor later.